service/header/exchange.go

```diff
@@ -0,0 +1,194 @@
+package header
+
+import (
+	"context"
+	"fmt"
+
+	header_pb "github.com/celestiaorg/celestia-node/service/header/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+var headerExchangeProtocolID = protocol.ID("expected-exchange")
+
+// exchange enables sending outbound ExtendedHeaderRequests as well as
+// handling inbound ExtendedHeaderRequests.
+type exchange struct {
+	host host.Host
+	// TODO @renaynay: post-Devnet, we need to remove reliance of Exchange on one bootstrap peer
+	// Ref https://github.com/celestiaorg/celestia-node/issues/172#issuecomment-964306823.
+	peer  *peer.AddrInfo
+	store Store
+}
+
+func newExchange(host host.Host, peer *peer.AddrInfo, store Store) *exchange {
+	ex := &exchange{
+		host:  host,
+		peer:  peer,
+		store: store,
+	}
+	host.SetStreamHandler(headerExchangeProtocolID, ex.requestHandler)
+	return ex
+}
+
+// requestHandler handles inbound ExtendedHeaderRequests.
+func (e *exchange) requestHandler(stream network.Stream) {
+	// unmarshal request
+	pbreq := new(header_pb.ExtendedHeaderRequest)
+	_, err := serde.Read(stream, pbreq)
```

Comment:
I'm surprised to see this used here. Looking at the git commit history: https://github.com/celestiaorg/celestia-node/pull/188/commits/7d6ef08c9344175fbe9be892f9765aad614683e5
it seems like this is only used to length-prefix the messages when writing and using that info when reading? Is that correct? I'm honestly not sure this justifies yet another dependency. I'd prefer to copy the necessary minimal logic into this repo instead. Core/tendermint used to provide a  helper for that (and it already is a dependency) but it got internalized: https://github.com/celestiaorg/celestia-core/tree/master/internal/libs/protoio

But also [another already used dependency](https://github.com/celestiaorg/celestia-node/blob/c2d6a3537259d910bee337017ab90d6d5e82662b/go.mod#L12) provides that same functionality: https://github.com/gogo/protobuf/blob/226206f39bd7276e88ec684ea0028c18ec2c91ae/io/varint.go#L44 and https://github.com/gogo/protobuf/blob/226206f39bd7276e88ec684ea0028c18ec2c91ae/io/varint.go#L94
This exists in gogo/proto as it is common to length prefix to read and write from streams in protobuf: https://developers.google.com/protocol-buffers/docs/techniques#streaming

Also, we know the max size of the extended header (and it's not really large).

Reply:
None of the mentioned utilities provide buffer pooling, but up to ofc...

Reply:
Yeah, I honetsly forgot about the buffer pooling. Maybe you are right and that justifies the additional dependency. Not sure. I would have literally copied that helper code into this repo and made it so that the max message is exactly the max extended header size instead. on another note, do you think it makes sense to let the user choose the max in the serde library instead?

Reply:
> Here, so far, I only see a very simple request of something that won't constantly be requested in large numbers.

It's header sync. Full nodes will serve millions of such msgs per day for light clients and other nodes.

Reply:
> On another note, do you think it makes sense to let the user choose the max in the serde library instead?
Yes, that is why there is a TODO saying so.

Reply:
>  Full nodes will serve thousands of such msgs per day for light clients and other nodes.

Yeah, not sure that really justifies a pool still. but all good. Let's revisit mem consumption and load once we have a more realistic network or a more realistic testing framework.